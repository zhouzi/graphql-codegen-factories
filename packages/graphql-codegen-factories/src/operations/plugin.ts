import { Kind, FragmentDefinitionNode, concatAST, DocumentNode } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesOperationsVisitor,
  FactoriesOperationsVisitorRawConfig,
} from "./FactoriesOperationsVisitor";

export const plugin: PluginFunction<
  FactoriesOperationsVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const allAst = concatAST(
    documents.map(({ document }) => document as DocumentNode)
  );
  const fragments = allAst.definitions.filter(
    (d) => d.kind === Kind.FRAGMENT_DEFINITION
  ) as FragmentDefinitionNode[];
  const allFragments: FragmentDefinitionNode[] = [
    ...fragments,
    // `externalFragments` config is passed by the near-operation-file preset.
    // It is an array of fragments declared outside of the operation file.
    ...(config.externalFragments || []).map(({ node }) => node),
  ];

  const visitor = new FactoriesOperationsVisitor(schema, allFragments, config);
  const content = (
    oldVisit(allAst, { leave: visitor }) as DocumentNode
  ).definitions
    .filter((definition) => typeof definition === "string")
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
