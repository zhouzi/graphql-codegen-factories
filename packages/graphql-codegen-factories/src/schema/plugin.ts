import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import { FactoriesBaseVisitorRawConfig } from "../FactoriesBaseVisitor";
import { FactoriesSchemaVisitor } from "./FactoriesSchemaVisitor";

export const plugin: PluginFunction<
  FactoriesBaseVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);

  const visitor = new FactoriesSchemaVisitor(schema, config);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
