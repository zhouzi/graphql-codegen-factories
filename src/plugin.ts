import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesVisitor,
  FactoriesVisitorRawConfig,
} from "./FactoriesVisitor";

export const plugin: PluginFunction<
  FactoriesVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const visitor = new FactoriesVisitor(schema, config);
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.config.typesPath
      ? [
          `import * as ${visitor.config.namespacedImportName} from '${visitor.config.typesPath}';\n`,
        ]
      : [],
    content,
  };
};
