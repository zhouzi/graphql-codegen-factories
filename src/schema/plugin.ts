import { printSchema, parse } from "graphql";
import {
  oldVisit,
  PluginFunction,
  Types,
} from "@graphql-codegen/plugin-helpers";
import {
  FactoriesSchemaVisitor,
  FactoriesSchemaVisitorRawConfig,
} from "./FactoriesSchemaVisitor";

export const plugin: PluginFunction<
  FactoriesSchemaVisitorRawConfig,
  Types.ComplexPluginOutput
> = (schema, documents, config) => {
  const visitor = new FactoriesSchemaVisitor(schema, config);
  const printedSchema = printSchema(schema);
  const astNode = parse(printedSchema);
  const content = oldVisit(astNode, { leave: visitor })
    .definitions.filter(Boolean)
    .join("\n");

  return {
    prepend: visitor.getImports(),
    content,
  };
};
