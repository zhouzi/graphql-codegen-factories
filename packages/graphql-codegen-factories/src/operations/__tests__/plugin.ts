import { buildSchema, parse, FragmentDefinitionNode, Kind } from "graphql";
import { plugin } from "../plugin";

describe("plugin", () => {
  it("should generate factory for a simple operation", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Mutation {
        createUser(username: String): User!
      }
    `);
    const ast = parse(/* GraphQL */ `
      mutation CreateUser($username: String) {
        createUser(username: $username) {
          id
          username
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "CreateUser.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should support aliases", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Mutation {
        createUser(username: String): User!
      }
    `);
    const ast = parse(/* GraphQL */ `
      mutation CreateUser($username: String) {
        createUser(username: $username) {
          id
          email: username
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "CreateUser.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should support fragments", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Query {
        me: User!
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMe {
        me {
          ...UserFragment
        }
      }
      fragment UserFragment on User {
        id
        username
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMe.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should merge fragments with the same type condition", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Query {
        me: User!
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMe {
        me {
          ...UserIDFragment
          ...UserUsernameFragment
        }
      }
      fragment UserIDFragment on User {
        id
      }
      fragment UserUsernameFragment on User {
        username
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMe.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should support inline fragments", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Query {
        me: User!
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMe {
        me {
          ... on User {
            id
            username
          }
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMe.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should merge inline fragments with the same type condition", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Query {
        me: User!
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMe {
        me {
          ... on User {
            id
          }
          ... on User {
            username
          }
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMe.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should unwrap inline fragments without a type condition", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Query {
        me: User!
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMe {
        me {
          ... {
            id
            ... {
              username
            }
          }
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMe.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should merge fragments and inline fragments with the same type condition", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Query {
        me: User!
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMe {
        me {
          ... on User {
            id
          }
          ...UserUsernameFragment
        }
      }
      fragment UserUsernameFragment on User {
        username
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMe.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should support external fragments", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Query {
        me: User!
      }
    `);

    const fragments = parse(/* GraphQL */ `
      fragment UserFragment on User {
        id
        username
      }
    `);

    const allFragments = fragments.definitions.filter(
      (d) => d.kind === Kind.FRAGMENT_DEFINITION
    ) as FragmentDefinitionNode[];

    const externalFragments = allFragments.map((frag) => ({
      isExternal: true,
      importFrom: frag.name.value,
      name: frag.name.value,
      onType: frag.typeCondition.name.value,
      node: frag,
    }));

    const ast = parse(/* GraphQL */ `
      query GetMe {
        me {
          ...UserFragment
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMe.graphql", document: ast }],
      {
        schemaFactoriesPath: "./factories",
        externalFragments,
      }
    );
    expect(output).toMatchSnapshot();
  });

  it("should support unnamed operations", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Query {
        me: User
      }
    `);
    const ast = parse(/* GraphQL */ `
      query {
        me {
          id
          username
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMe.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should support lists", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
      }

      type Query {
        users: [User!]!
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetUsers {
        users {
          id
          username
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetUsers.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should support nested selections", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type User {
        id: ID!
        username: String!
        followers: [User]
      }

      type Query {
        me: User
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMe {
        me {
          id
          username
          followers {
            id
          }
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMe.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should support unions", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type ImageDimensions {
        width: Int!
        height: Int!
      }

      type Image {
        src: String!
        dimensions: ImageDimensions
      }

      type Video {
        href: String!
        dimensions: ImageDimensions
      }

      union Media = Image | Video

      type Query {
        medias: [Media!]!
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMedias {
        medias {
          ... on Image {
            src
            dimensions {
              width
            }
          }
          ... on Video {
            href
            dimensions {
              height
            }
          }
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMedias.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should add interface's selections to the matching types", async () => {
    const schema = buildSchema(/* GraphQL */ `
      interface File {
        path: String!
      }

      type Image implements File {
        path: String!
        width: Int!
        height: Int!
      }

      type Audio implements File {
        path: String!
        length: Int!
      }

      interface Streamable {
        url: String!
      }

      type Video implements Streamable {
        url: String!
        length: Int!
      }

      union Media = Image | Audio | Video

      type Query {
        medias: [Media!]!
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMedias {
        medias {
          ... on File {
            path
          }
          ... on Streamable {
            url
          }
          ... on Image {
            width
          }
          ... on Audio {
            length
          }
          ... on Video {
            length
          }
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMedias.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should dedupe fields", async () => {
    const schema = buildSchema(/* GraphQL */ `
      interface Node {
        id: ID!
      }

      type User implements Node {
        id: ID!
        username: String!
      }

      type Admin implements Node {
        id: ID!
        canDeleteUser: Boolean!
      }

      union Me = User | Admin

      type Query {
        me: Me
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMe {
        me {
          ... on Node {
            id
          }
          ... on User {
            ...UserFragment
            id
            userId: id
            username
          }
          ... on Admin {
            id
          }
        }
      }
      fragment UserFragment on User {
        id
        username
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMe.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });

  it("should generate union factory even when querying one type from the union", async () => {
    const schema = buildSchema(/* GraphQL */ `
      type Image {
        width: Int!
        height: Int!
      }

      type Audio {
        length: Int!
      }

      union Media = Image | Audio

      type Query {
        medias: [Media!]!
      }
    `);
    const ast = parse(/* GraphQL */ `
      query GetMedias {
        medias {
          ... on Audio {
            length
          }
        }
      }
    `);

    const output = await plugin(
      schema,
      [{ location: "GetMedias.graphql", document: ast }],
      { schemaFactoriesPath: "./factories" }
    );
    expect(output).toMatchSnapshot();
  });
});
