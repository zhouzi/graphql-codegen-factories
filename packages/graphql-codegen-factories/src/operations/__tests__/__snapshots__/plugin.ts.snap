// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin should generate factory for a simple operation 1`] = `
Object {
  "content": "export function createCreateUserMutationMock(props: Partial<CreateUserMutation>): CreateUserMutation {
  switch(props.__typename) {
    case \\"Mutation\\": {
      const {  } = factories.createMutationMock({  });
      return { createUser: createCreateUserMutationMock_createUser({}), ...props };
    }
    case undefined:
    default:
      return createCreateUserMutationMock({ ...props, __typename: \\"Mutation\\" });
  }
}
export function createCreateUserMutationMock_createUser(props: Partial<CreateUserMutation[\\"createUser\\"]>): CreateUserMutation[\\"createUser\\"] {
  switch(props.__typename) {
    case \\"User\\": {
      const { id, username } = factories.createUserMock({ id: props.id, username: props.username });
      return { id, username, ...props };
    }
    case undefined:
    default:
      return createCreateUserMutationMock_createUser({ ...props, __typename: \\"User\\" });
  }
}",
  "prepend": Array [
    "import * as factories from \\"./factories\\";",
  ],
}
`;

exports[`plugin should support aliases 1`] = `
Object {
  "content": "export function createCreateUserMutationMock(props: Partial<CreateUserMutation>): CreateUserMutation {
  switch(props.__typename) {
    case \\"Mutation\\": {
      const {  } = factories.createMutationMock({  });
      return { createUser: createCreateUserMutationMock_createUser({}), ...props };
    }
    case undefined:
    default:
      return createCreateUserMutationMock({ ...props, __typename: \\"Mutation\\" });
  }
}
export function createCreateUserMutationMock_createUser(props: Partial<CreateUserMutation[\\"createUser\\"]>): CreateUserMutation[\\"createUser\\"] {
  switch(props.__typename) {
    case \\"User\\": {
      const { id, username: email } = factories.createUserMock({ id: props.id, username: props.email });
      return { id, email, ...props };
    }
    case undefined:
    default:
      return createCreateUserMutationMock_createUser({ ...props, __typename: \\"User\\" });
  }
}",
  "prepend": Array [
    "import * as factories from \\"./factories\\";",
  ],
}
`;

exports[`plugin should support external fragments 1`] = `
Object {
  "content": "export function createGetMeQueryMock(props: Partial<GetMeQuery>): GetMeQuery {
  switch(props.__typename) {
    case \\"Query\\": {
      const {  } = factories.createQueryMock({  });
      return { me: createGetMeQueryMock_me({}), ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock({ ...props, __typename: \\"Query\\" });
  }
}
export function createGetMeQueryMock_me(props: Partial<GetMeQuery[\\"me\\"]>): GetMeQuery[\\"me\\"] {
  switch(props.__typename) {
    case \\"User\\": {
      const {  } = factories.createUserMock({  });
      return { ...createGetMeQueryMock_me_User({}), ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock_me({ ...props, __typename: \\"User\\" });
  }
}
export function createGetMeQueryMock_me_User(props: Partial<Extract<GetMeQuery[\\"me\\"], { __typename: \\"User\\" }>>): Extract<GetMeQuery[\\"me\\"], { __typename: \\"User\\" }> {
  switch(props.__typename) {
    case \\"User\\": {
      const { id, username } = factories.createUserMock({ id: props.id, username: props.username });
      return { id, username, ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock_me_User({ ...props, __typename: \\"User\\" });
  }
}",
  "prepend": Array [
    "import * as factories from \\"./factories\\";",
  ],
}
`;

exports[`plugin should support fragments 1`] = `
Object {
  "content": "export function createGetMeQueryMock(props: Partial<GetMeQuery>): GetMeQuery {
  switch(props.__typename) {
    case \\"Query\\": {
      const {  } = factories.createQueryMock({  });
      return { me: createGetMeQueryMock_me({}), ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock({ ...props, __typename: \\"Query\\" });
  }
}
export function createGetMeQueryMock_me(props: Partial<GetMeQuery[\\"me\\"]>): GetMeQuery[\\"me\\"] {
  switch(props.__typename) {
    case \\"User\\": {
      const {  } = factories.createUserMock({  });
      return { ...createGetMeQueryMock_me_User({}), ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock_me({ ...props, __typename: \\"User\\" });
  }
}
export function createGetMeQueryMock_me_User(props: Partial<Extract<GetMeQuery[\\"me\\"], { __typename: \\"User\\" }>>): Extract<GetMeQuery[\\"me\\"], { __typename: \\"User\\" }> {
  switch(props.__typename) {
    case \\"User\\": {
      const { id, username } = factories.createUserMock({ id: props.id, username: props.username });
      return { id, username, ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock_me_User({ ...props, __typename: \\"User\\" });
  }
}",
  "prepend": Array [
    "import * as factories from \\"./factories\\";",
  ],
}
`;

exports[`plugin should support inline fragments 1`] = `
Object {
  "content": "export function createGetMeQueryMock(props: Partial<GetMeQuery>): GetMeQuery {
  switch(props.__typename) {
    case \\"Query\\": {
      const {  } = factories.createQueryMock({  });
      return { me: createGetMeQueryMock_me({}), ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock({ ...props, __typename: \\"Query\\" });
  }
}
export function createGetMeQueryMock_me(props: Partial<GetMeQuery[\\"me\\"]>): GetMeQuery[\\"me\\"] {
  switch(props.__typename) {
    case \\"User\\": {
      const {  } = factories.createUserMock({  });
      return { ...createGetMeQueryMock_me_User({}), ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock_me({ ...props, __typename: \\"User\\" });
  }
}
export function createGetMeQueryMock_me_User(props: Partial<Extract<GetMeQuery[\\"me\\"], { __typename: \\"User\\" }>>): Extract<GetMeQuery[\\"me\\"], { __typename: \\"User\\" }> {
  switch(props.__typename) {
    case \\"User\\": {
      const { id, username } = factories.createUserMock({ id: props.id, username: props.username });
      return { id, username, ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock_me_User({ ...props, __typename: \\"User\\" });
  }
}",
  "prepend": Array [
    "import * as factories from \\"./factories\\";",
  ],
}
`;

exports[`plugin should support lists 1`] = `
Object {
  "content": "export function createGetUsersQueryMock(props: Partial<GetUsersQuery>): GetUsersQuery {
  switch(props.__typename) {
    case \\"Query\\": {
      const {  } = factories.createQueryMock({  });
      return { users: [], ...props };
    }
    case undefined:
    default:
      return createGetUsersQueryMock({ ...props, __typename: \\"Query\\" });
  }
}
export function createGetUsersQueryMock_users(props: Partial<GetUsersQuery[\\"users\\"][number]>): GetUsersQuery[\\"users\\"][number] {
  switch(props.__typename) {
    case \\"User\\": {
      const { id, username } = factories.createUserMock({ id: props.id, username: props.username });
      return { id, username, ...props };
    }
    case undefined:
    default:
      return createGetUsersQueryMock_users({ ...props, __typename: \\"User\\" });
  }
}",
  "prepend": Array [
    "import * as factories from \\"./factories\\";",
  ],
}
`;

exports[`plugin should support nested selections 1`] = `
Object {
  "content": "export function createGetMeQueryMock(props: Partial<GetMeQuery>): GetMeQuery {
  switch(props.__typename) {
    case \\"Query\\": {
      const {  } = factories.createQueryMock({  });
      return { me: null, ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock({ ...props, __typename: \\"Query\\" });
  }
}
export function createGetMeQueryMock_me(props: Partial<NonNull<GetMeQuery[\\"me\\"]>>): NonNull<GetMeQuery[\\"me\\"]> {
  switch(props.__typename) {
    case \\"User\\": {
      const { id, username } = factories.createUserMock({ id: props.id, username: props.username });
      return { id, username, followers: null, ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock_me({ ...props, __typename: \\"User\\" });
  }
}
export function createGetMeQueryMock_me_followers(props: Partial<NonNull<NonNull<NonNull<GetMeQuery[\\"me\\"]>[\\"followers\\"]>[number]>>): NonNull<NonNull<NonNull<GetMeQuery[\\"me\\"]>[\\"followers\\"]>[number]> {
  switch(props.__typename) {
    case \\"User\\": {
      const { id } = factories.createUserMock({ id: props.id });
      return { id, ...props };
    }
    case undefined:
    default:
      return createGetMeQueryMock_me_followers({ ...props, __typename: \\"User\\" });
  }
}",
  "prepend": Array [
    "import * as factories from \\"./factories\\";",
  ],
}
`;

exports[`plugin should support unions 1`] = `
Object {
  "content": "export function createGetMediasQueryMock(props: Partial<GetMediasQuery>): GetMediasQuery {
  switch(props.__typename) {
    case \\"Query\\": {
      const {  } = factories.createQueryMock({  });
      return { medias: [], ...props };
    }
    case undefined:
    default:
      return createGetMediasQueryMock({ ...props, __typename: \\"Query\\" });
  }
}
export function createGetMediasQueryMock_medias(props: Partial<GetMediasQuery[\\"medias\\"][number]>): GetMediasQuery[\\"medias\\"][number] {
  switch(props.__typename) {
    case \\"Image\\": {
      const {  } = factories.createImageMock({  });
      return { ...createGetMediasQueryMock_medias_Image({}), ...props };
    }
    case \\"Video\\": {
      const {  } = factories.createVideoMock({  });
      return { ...createGetMediasQueryMock_medias_Video({}), ...props };
    }
    case undefined:
    default:
      return createGetMediasQueryMock_medias({ ...props, __typename: \\"Image\\" });
  }
}
export function createGetMediasQueryMock_medias_Image(props: Partial<Extract<GetMediasQuery[\\"medias\\"][number], { __typename: \\"Image\\" }>>): Extract<GetMediasQuery[\\"medias\\"][number], { __typename: \\"Image\\" }> {
  switch(props.__typename) {
    case \\"Image\\": {
      const { src } = factories.createImageMock({ src: props.src });
      return { src, dimensions: null, ...props };
    }
    case undefined:
    default:
      return createGetMediasQueryMock_medias_Image({ ...props, __typename: \\"Image\\" });
  }
}
export function createGetMediasQueryMock_medias_Image_dimensions(props: Partial<NonNull<Extract<GetMediasQuery[\\"medias\\"][number], { __typename: \\"Image\\" }>[\\"dimensions\\"]>>): NonNull<Extract<GetMediasQuery[\\"medias\\"][number], { __typename: \\"Image\\" }>[\\"dimensions\\"]> {
  switch(props.__typename) {
    case \\"ImageDimensions\\": {
      const { width } = factories.createImageDimensionsMock({ width: props.width });
      return { width, ...props };
    }
    case undefined:
    default:
      return createGetMediasQueryMock_medias_Image_dimensions({ ...props, __typename: \\"ImageDimensions\\" });
  }
}
export function createGetMediasQueryMock_medias_Video(props: Partial<Extract<GetMediasQuery[\\"medias\\"][number], { __typename: \\"Video\\" }>>): Extract<GetMediasQuery[\\"medias\\"][number], { __typename: \\"Video\\" }> {
  switch(props.__typename) {
    case \\"Video\\": {
      const { href } = factories.createVideoMock({ href: props.href });
      return { href, dimensions: null, ...props };
    }
    case undefined:
    default:
      return createGetMediasQueryMock_medias_Video({ ...props, __typename: \\"Video\\" });
  }
}
export function createGetMediasQueryMock_medias_Video_dimensions(props: Partial<NonNull<Extract<GetMediasQuery[\\"medias\\"][number], { __typename: \\"Video\\" }>[\\"dimensions\\"]>>): NonNull<Extract<GetMediasQuery[\\"medias\\"][number], { __typename: \\"Video\\" }>[\\"dimensions\\"]> {
  switch(props.__typename) {
    case \\"ImageDimensions\\": {
      const { height } = factories.createImageDimensionsMock({ height: props.height });
      return { height, ...props };
    }
    case undefined:
    default:
      return createGetMediasQueryMock_medias_Video_dimensions({ ...props, __typename: \\"ImageDimensions\\" });
  }
}",
  "prepend": Array [
    "import * as factories from \\"./factories\\";",
  ],
}
`;

exports[`plugin should support unnamed operations 1`] = `
Object {
  "content": "export function createUnnamed_1_QueryMock(props: Partial<Unnamed_1_Query>): Unnamed_1_Query {
  switch(props.__typename) {
    case \\"Query\\": {
      const {  } = factories.createQueryMock({  });
      return { me: null, ...props };
    }
    case undefined:
    default:
      return createUnnamed_1_QueryMock({ ...props, __typename: \\"Query\\" });
  }
}
export function createUnnamed_1_QueryMock_me(props: Partial<NonNull<Unnamed_1_Query[\\"me\\"]>>): NonNull<Unnamed_1_Query[\\"me\\"]> {
  switch(props.__typename) {
    case \\"User\\": {
      const { id, username } = factories.createUserMock({ id: props.id, username: props.username });
      return { id, username, ...props };
    }
    case undefined:
    default:
      return createUnnamed_1_QueryMock_me({ ...props, __typename: \\"User\\" });
  }
}",
  "prepend": Array [
    "import * as factories from \\"./factories\\";",
  ],
}
`;
